### PR TITLE
MESH-1675: Don't check perms in `Identity::leave_identity_as_key` + Remove some extrinsics

### PR DESCRIPTION
**Pull request by Centril**
### modified logic

- Do not check permissions in `Identity::leave_identity_as_key`.

### modified api

- Removed extrinsics `Identity::{join, leave}_identity_as_identity` and `Identity::revoke_offchain_authorization`.
- Removed event `Identity.ForwardedCall`.
- Removed errors `{Permissions, Identity}.RecursionNotAllowed`

### CODE REVIEW

**Review by Neopallium — CHANGES_REQUESTED** [review]

File: pallets/identity/src/lib.rs [hunk 1]
```diff
@@ -439,14 +439,14 @@ decl_module! {
...
+            Self::leave_identity(Signatory::Account(sender), did)
         }
 
         /// Leave an identity as a secondary identity.
         #[weight = <T as Config>::WeightInfo::leave_identity_as_identity()]
         pub fn leave_identity_as_identity(origin, did: IdentityId) -> DispatchResult {
-            let sender_did = Self::ensure_perms(origin)?;
+            let (_, sender_did) = Self::ensure_did(origin)?;
             Self::leave_identity(Signatory::from(sender_did), did)
         }
```
[the remark this answers is not in the document]
Comment (F-OBrien) [reply pallets/identity/src/lib.rs:451]: My opinion (like it matters 😄), which I think was the intention of this change, is that a signing identity should not need permission to leave another identity. I did notice this previously but never said anything.

I do however think there should be a `is_primary_key` check applied here so only the primary key of the identity leaving can call it.

[the remark this answers is not in the document]
Comment (Neopallium) [reply pallets/identity/src/lib.rs:451]: The signing identity doesn't need permissions to make this call and leave another identity.

[the remark this answers is not in the document]
Comment (Centril) [reply pallets/identity/src/lib.rs:451]: Great catch! Fixing :)

[the remark this answers is not in the document]
Comment (F-OBrien) [reply pallets/identity/src/lib.rs:451]: Not exactly related to this PR but something similar around permissions to leave a roll. Someone who is added as an attestation provider has no ability to remove themselves if they wish to "quit". `removeDefaultTrustedClaimIssuer` fails if called by the Claim Issuer and there is no other function to leave

File: pallets/identity/src/lib.rs [hunk 2]
```diff
@@ -2138,64 +2137,57 @@ impl<T: Config> CheckAccountCallPermissions<T::AccountId> for Module<T> {
...
-
-        if !<KeyToIdentityIds<T>>::contains_key(who) {
-            // Caller has no Identity
-            return None;
-        }
+        // `who` is the original origin (signer) of the original extrinsic.
+        // `context::current_identity` (CI) is the target DID for which the calling key must have permissions.
+        // If `who`'s identity does not match CI we have a forwarded call.
+        // In this case, the key for which we check permissions is `who`'s DID rather than `who`.
+        // This assumes that recursive forwarded calls, are not allowed, as enforced in `forwarded_call`.
```
Comment (Neopallium) [inline pallets/identity/src/lib.rs:2144]: We don't have a `forwarded_call`.  I think this comment is referring to some older logic?

The only reason I see `who`'s identity not matching `context::current_identity` is from `CddAndFeeDetails` when a different identity pays the transaction fee (for accepting an authorization).

Comment (Neopallium) [reply pallets/identity/src/lib.rs:2144]: Also we have a `ForwardedCall` event that is not created anywhere.

Comment (F-OBrien) [reply pallets/identity/src/lib.rs:2144]: forwarded calls were removed in #990

**Comment by F-OBrien** [comment]
`pub fn get_identity_of_signatory` appears to be unused from the identity pallet. If it has no purpose you may want to removed before mainnet.

File: pallets/identity/src/lib.rs [hunk 3]
```diff
@@ -2123,64 +2123,57 @@ impl<T: Config> CheckAccountCallPermissions<T::AccountId> for Module<T> {
...
-                    .map(|sk| AccountCallPermissionsData {
-                        primary_did: target_did,
-                        secondary_key: Some(sk),
-                    })
-            };
+        let who_sk = if is_direct_call {
+            // Direct call. We'll check the permissions of `who` in `target_did`.
+            Signatory::Account(who.clone())
+        } else {
+            // Forwarded call. We'll check the permissions of `who`'s DID in `target_did`.
```
Comment (adamdossa) [inline pallets/identity/src/lib.rs:2163]: Wondering if we really need this, or whether `!is_direct_call` is an error condition now that we don't have forwarded calls.

File: pallets/identity/src/lib.rs [hunk 4]
```diff
@@ -439,14 +439,17 @@ decl_module! {
...
         }
 
         /// Leave an identity as a secondary identity.
         #[weight = <T as Config>::WeightInfo::leave_identity_as_identity()]
         pub fn leave_identity_as_identity(origin, did: IdentityId) -> DispatchResult {
-            let sender_did = Self::ensure_perms(origin)?;
+            let (sender, sender_did) = Self::ensure_did(origin)?;
+            if sender != DidRecords::<T>::get(sender_did).primary_key {
+                CallPermissions::<T>::ensure_call_permissions(&sender)?;
+            }
```
Comment (Neopallium) [inline pallets/identity/src/lib.rs:452]: This change is still not needed.  The new tests pass with the original permissions check:
`let sender_did = Self::ensure_perms(origin)?;`

File: pallets/identity/src/lib.rs [hunk 5]
```diff
@@ -75,13 +75,12 @@
 //! - `add_authorization` - Adds an authorization.
 //! - `remove_authorization` - Removes an authorization.
 //! - `add_secondary_keys_with_authorization` - Adds secondary keys to target identity `id`.
-//! - `revoke_offchain_authorization` - Revokes the `auth` off-chain authorization of `signer`.
```
Comment (F-OBrien) [inline pallets/identity/src/lib.rs:78]: "//! - `join_identity_as_identity` - Join an identity as a secondary identity." also needs to be removed from the comment above here

File: pallets/identity/src/lib.rs [hunk 6]
```diff
@@ -875,43 +805,28 @@ impl<T: Config> Module<T> {
...
+            // Charge the protocol fee after all checks.
+            T::ProtocolFee::charge_fee(ProtocolOp::IdentityAddSecondaryKeysWithAuthorization)?;
+            // Update current did of the transaction to the newly joined did.
+            // This comes handy when someone uses a batch transaction to leave their identity,
+            // join another identity, and then do something as the new identity.
+            T::CddHandler::set_current_identity(&target_did);
+
+            Self::link_account_key_to_did(&key, target_did);
+
+            Self::unsafe_join_identity(target_did, permissions, key);
```
[the remark this answers is not in the document]
Comment (Centril) [reply pallets/identity/src/lib.rs:829]: Could we do this in a follow-up? The scope of this PR has already grown a lot.

**Review by CJP10 — APPROVED** [review]

File: pallets/identity/src/lib.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (Neopallium) [reply pallets/identity/src/lib.rs:829]: It is a small change.  Why do another PR just to move one line of code and delete another line.

[the remark this answers is not in the document]
Comment (Centril) [reply pallets/identity/src/lib.rs:829]: Alright, I'll fix it here.

**Review by Neopallium — APPROVED** [review]

**Pull request merged**